libzeth/serialization/stream_utils.hpp

```diff
@@ -0,0 +1,103 @@
+// Copyright (c) 2015-2021 Clearmatics Technologies Ltd
+//
+// SPDX-License-Identifier: LGPL-3.0+
+
+#ifndef __ZETH_SERIALIZATION_STREAM_UTILS_HPP__
+#define __ZETH_SERIALIZATION_STREAM_UTILS_HPP__
+
+#include "libzeth/core/include_libsnark.hpp"
+
+#include <iostream>
+#include <type_traits>
+
+namespace libzeth
+{
+
+/// Statically derive the type of the element contained in a (vector-like)
+/// collection.
+template<typename CollectionT>
+using MemberT =
+    typename std::decay<decltype((*(CollectionT *)nullptr)[0])>::type;
```

Comment:
Same question as above here re: the decay types. Wouldn't it be sufficient to simply get the type of a member of the collection by using `decltype`? Why exactly do we need the surrounding `decay::type` here?

Reply:
The index operator returns a reference or const reference, whereas we need `MemberT` to be the type of the underlying object in the collection.  As above, `std::decay` strips away any const / ref parts of the type.